Cauli/CauliURLProtocol/InterceptingInputStream.swift

```diff
@@ -0,0 +1,79 @@
+//
+//  InterceptingInputStream.swift
+//  Cauliframework
+//
+//  Created by Cornelius Horstmann on 19.02.19.
+//
+
+import Foundation
+
+internal protocol InterceptingInputStreamDelegate {
+    func didRead(_ buffer: UnsafeMutablePointer<UInt8>, maxLength len: Int)
+}
+
+internal class InterceptingInputStream: InputStream {
+    
+    let parentStream: InputStream
+    
+    var interceptingDelegate: InterceptingInputStreamDelegate?
+    override var delegate: StreamDelegate? {
+        get {
+            return parentStream.delegate
+        }
+        set {
+            parentStream.delegate = delegate
+        }
+    }
+    
+    override var streamStatus: Stream.Status {
+        return parentStream.streamStatus
+    }
+    
+    override var streamError: Error? {
+        return parentStream.streamError
+    }
+    
+    init(parentStream: InputStream) {
+        self.parentStream = parentStream
+        super.init(data: Data())
```

Comment:
This is the one downside of this `InterceptingInputStream`. We have to call super.init with actual Data. It works this way, because we are passing all actual functions to the parentStream but there are risks:

* All private/internal functions are not intercepted 
* If Apple ever adds functions we will not notice them

Reply:
Let's live with the risks. Maybe we can mark that feature as kind of experimental in our documentation to point out, there might be issues with this. Also we could add an annotation, so that the compiler will tell us for iOS 12 that there could be changes. Not sure.

But honestly. With each new iOS version we should check the core functionality. Without checking we're also not getting any knowledge about changes in the URLprotocol for instance.

Reply:
What do you think about adding a setting to the configuration, explaining that this is an experimental feature and disable it per default? I feel a bit weird thinking about enabling an experimental feature per default.

Reply:
I just digged into this a little more and actually it's not possible to do this as a configuration on Cauli level, since (in this implementation) all the logic is on the CauliURLProtocol level. But: Why don't we just implement a Floret, that will simply modify the Request? I will try to implement that, but I don't see any reason why it shouldn't work.